lib/ts/recipe/thirdParty/utils.ts

```diff
@@ -0,0 +1,186 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { NormalisedAppinfo } from "../../types";
+import { validateTheStructureOfUserInput } from "../../utils";
+import Recipe from "./recipe";
+import STError from "./error";
+import { TypeInput as TypeNormalisedInputEmailVerification } from "../emailverification/types";
+import {
+    InputSchema,
+    TypeInput,
+    TypeInputEmailVerificationFeature,
+    TypeNormalisedInput,
+    TypeNormalisedInputSignOutFeature,
+    TypeInputSignOutFeature,
+    TypeNormalisedInputSignInAndUp,
+    TypeInputSignInAndUp,
+} from "./types";
+
+export function validateAndNormaliseUserInput(
+    recipeInstance: Recipe,
+    appInfo: NormalisedAppinfo,
+    config: TypeInput
+): TypeNormalisedInput {
+    validateTheStructureOfUserInput(config, InputSchema, "thirdparty recipe", recipeInstance.getRecipeId());
+    let emailVerificationFeature = validateAndNormaliseEmailVerificationConfig(
+        recipeInstance,
+        appInfo,
+        config.emailVerificationFeature
+    );
+
+    let signInAndUpFeature = validateAndNormaliseSignInAndUpConfig(recipeInstance, appInfo, config.signInAndUpFeature);
+
+    let signOutFeature = validateAndNormaliseSignOutConfig(recipeInstance, appInfo, config.signOutFeature);
+
+    return {
+        emailVerificationFeature,
+        signOutFeature,
+        signInAndUpFeature,
+    };
+}
+
+function validateAndNormaliseSignInAndUpConfig(
+    recipeInstance: Recipe,
+    appInfo: NormalisedAppinfo,
+    config?: TypeInputSignInAndUp
+): TypeNormalisedInputSignInAndUp {
+    if (config === undefined) {
+        throw new STError(
+            {
+                type: "BAD_INPUT_ERROR",
+                message: "thirdparty recipe requires atleast signInAndUpFeature config to be passed",
+            },
+            recipeInstance.getRecipeId()
+        );
+    }
+    let disableDefaultImplementation =
+        config.disableDefaultImplementation === undefined ? false : config.disableDefaultImplementation;
+
+    let handlePostSignUpIn = config.handlePostSignUpIn;
+
+    if (handlePostSignUpIn === undefined) {
```

Comment:
Not true.. This is optional. If the user has not provided it, the default function should do nothing.